File: apps/web/src/ai/editor-adapter/audio-plan.ts
import { z } from "zod";
import type { AudioCatalogV1, AudioPlanV1 } from "./contracts";

const audioLicenseSchema = z.strictObject({
	licenseId: z.string(),
	licenseName: z.string(),
	sourceUrl: z.string(),
	licenseUrl: z.string(),
	creator: z.string(),
	attributionRequired: z.boolean(),
	attributionText: z.string().nullable(),
	commercialUseAllowed: z.boolean(),
	derivativeUseAllowed: z.boolean(),
	reviewStatus: z.enum(["verified", "review-needed"]),
	reviewedAt: z.string().nullable(),
});

const audioCatalogAssetSchema = z.strictObject({
	id: z.string(),
	mediaId: z.string().nullable(),
	kind: z.enum(["music", "sound-effect", "transition"]),
	durationTicks: z.number().int(),
	license: audioLicenseSchema,
});

const audioCatalogSchema: z.ZodType<AudioCatalogV1> = z.strictObject({
	schemaVersion: z.literal("1.0"),
	catalogId: z.string(),
	revision: z.number().int().nonnegative(),
	assets: z.array(audioCatalogAssetSchema),
});

const audioDuckingSchema = z.strictObject({
	enabled: z.boolean(),
	targetGainDb: z.number(),
	attackTicks: z.number().int(),
	releaseTicks: z.number().int(),
	dialogueTrackIds: z.array(z.string()),
});

const insertCatalogAudioSchema = z.strictObject({
	type: z.literal("insert_catalog_audio"),
	operationId: z.string(),
	resultElementId: z.string(),
	catalogAssetId: z.string(),
	targetTrackId: z.string(),
	timelineStartTicks: z.number().int(),
	sourceStartTicks: z.number().int(),
	durationTicks: z.number().int(),
	gainDb: z.number(),
	fadeInTicks: z.number().int(),
	fadeOutTicks: z.number().int(),
	ducking: audioDuckingSchema.nullable(),
});

const audioPlanSchema: z.ZodType<AudioPlanV1> = z.strictObject({
	schemaVersion: z.literal("1.0"),
	planId: z.string(),
	idempotencyKey: z.string(),
	projectId: z.string(),
	baseProjectRevision: z.number().int().nonnegative(),
	baseTimelineHash: z.string(),
	catalogId: z.string(),
	catalogRevision: z.number().int().nonnegative(),
	operations: z.array(insertCatalogAudioSchema).min(1).max(200),
});

function parseJsonInput({ input }: { input: string | unknown }): unknown {
	if (typeof input !== "string") return input;
	const fenced = input.match(/```(?:json)?\s*([\s\S]*?)```/i)?.[1];
	return JSON.parse((fenced ?? input).trim());
}

/** Strictly parses GPT output before Rust performs project/catalog validation. */
export function parseAudioPlanV1({
	input,
}: {
	input: string | unknown;
}): AudioPlanV1 {
	const result = audioPlanSchema.safeParse(parseJsonInput({ input }));
	if (!result.success) {
		throw new Error("JSON bukan AudioPlanV1 yang valid.", {
			cause: result.error,
		});
	}
	return result.data;
}

/** Strictly parses catalog metadata supplied to the GPT bridge and validator. */
export function parseAudioCatalogV1({
	input,
}: {
	input: string | unknown;
}): AudioCatalogV1 {
	const result = audioCatalogSchema.safeParse(parseJsonInput({ input }));
	if (!result.success) {
		throw new Error("JSON bukan AudioCatalogV1 yang valid.", {
			cause: result.error,
		});
	}
	return result.data;
}

File: rust/crates/editor-contracts/src/audio_plan.rs
use serde::{Deserialize, Serialize};

pub const AUDIO_PLAN_SCHEMA_VERSION: &str = "1.0";
pub const AUDIO_CATALOG_SCHEMA_VERSION: &str = "1.0";
pub const MAX_AUDIO_OPERATIONS: usize = 200;
pub const MIN_AUDIO_GAIN_DB: f64 = -60.0;
pub const MAX_AUDIO_GAIN_DB: f64 = 12.0;

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AudioCatalogV1 {
    pub schema_version: String,
    pub catalog_id: String,
    pub revision: u64,
    pub assets: Vec<AudioCatalogAssetV1>,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum AudioCatalogAssetKindV1 {
    Music,
    #[serde(rename = "sound-effect")]
    SoundEffect,
    Transition,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AudioCatalogAssetV1 {
    pub id: String,
    /// ID of the downloaded/cached audio asset in the project snapshot.
    /// It may be absent during preflight and is required before apply.
    pub media_id: Option<String>,
    pub kind: AudioCatalogAssetKindV1,
    pub duration_ticks: i64,
    pub license: AudioLicenseV1,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "snake_case")]
pub enum AudioPlanValidationPhaseV1 {
    /// Validates GPT output and catalog/license references before any download.
    Preflight,
    /// Additionally requires every selected asset to resolve to project audio media.
    ReadyToApply,
}

#[derive(Clone, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AudioLicenseV1 {
    /// Stable license identifier (for example `CC0-1.0` or a provider license ID).
    pub license_id: String,
    pub license_name: String,
    pub source_url: String,
    pub license_url: String,
    pub creator: String,
    pub attribution_required: bool,
    pub attribution_text: Option<String>,
    pub commercial_use_allowed: bool,
    pub derivative_use_allowed: bool,
    pub review_status: AudioLicenseReviewStatusV1,
    pub reviewed_at: Option<String>,
}

#[derive(Clone, Copy, Debug, Deserialize, Serialize, Eq, PartialEq)]
#[serde(rename_all = "kebab-case")]
pub enum AudioLicenseReviewStatusV1 {
    Verified,
    ReviewNeeded,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AudioPlanV1 {
    pub schema_version: String,
    pub plan_id: String,
    pub idempotency_key: String,
    pub project_id: String,
    pub base_project_revision: u64,
    pub base_timeline_hash: String,
    pub catalog_id: String,
    pub catalog_revision: u64,
    pub operations: Vec<AudioOperationV1>,
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(
    tag = "type",
    rename_all = "snake_case",
    rename_all_fields = "camelCase",
    deny_unknown_fields
)]
pub enum AudioOperationV1 {
    InsertCatalogAudio {
        operation_id: String,
        result_element_id: String,
        catalog_asset_id: String,
        target_track_id: String,
        timeline_start_ticks: i64,
        source_start_ticks: i64,
        duration_ticks: i64,
        gain_db: f64,
        fade_in_ticks: i64,
        fade_out_ticks: i64,
        ducking: Option<AudioDuckingV1>,
    },
}

impl AudioOperationV1 {
    pub fn operation_id(&self) -> &str {
        match self {
            Self::InsertCatalogAudio { operation_id, .. } => operation_id,
        }
    }
}

#[derive(Clone, Debug, Deserialize, Serialize, PartialEq)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AudioDuckingV1 {
    pub enabled: bool,
    /// Gain applied while dialogue is active. Must not amplify the inserted audio.
    pub target_gain_db: f64,
    pub attack_ticks: i64,
    pub release_ticks: i64,
    /// Audio tracks used as the dialogue sidechain. Empty is invalid when enabled.
    pub dialogue_track_ids: Vec<String>,
}
